source/modulo_core/src/Cell.cpp

```diff
@@ -419,6 +425,28 @@ bool Cell::on_cleanup() {
 
 rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn Cell::on_shutdown(const rclcpp_lifecycle::State& state) {
   RCUTILS_LOG_INFO_NAMED(get_name(), "on_shutdown() is called from state %s.", state.label().c_str());
+  uint8_t current_state = state.id();
+  // if the node is already shutdown just return success
+  if (state.id() == lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED) {
```

Comment:
for consistency, since you copy this value to `current_state`, I would suggest you also use it here
```suggestion
  if (current_state == lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED) {
```